### ISSUE TITLE
Moderately slow query on ALL_ARGUMENTS due to EXTENDED DATA LINK FULL in OracleTableDefinition

### ISSUE THREAD

**Issue by lukaseder** [jOOQ/jOOQ#14021]
While investigating another excessively slow query (https://github.com/jOOQ/jOOQ/issues/14014), I've noticed a moderately slow query that may prove to be an issue for larger schemas:

```sql
select 
  sql_id,
  child_number,
  elapsed_time / 1000000.0 as et, 
  executions as ex,
  cast(elapsed_time / 1000000.0 / nullif(executions, 0) as number(20, 3)) as tpe,
  sql_text
  -- s.*
from v$sql s
where 1 = 1
and module = 'JDBC Thin Client'
order by elapsed_time desc
;
```

The top 3

```
23mrzz17v4hhq	0	26.838084	438	0.061	select a.IN_OUT, a.ARGUMENT_NAME, a.DATA_TYPE, a.CHAR_LENGTH, a.DATA_PRECISION, a.DATA_SCALE, case when a.DATA_TYPE = 'REF CURSOR' then null when (a.DATA_TYPE = 'PL/SQL RECORD' and 1 = 0) then a.OWNER else a.TYPE_OWNER end TYPE_OWNER, case when a.DATA_TYPE = 'REF CURSOR' then null when (a.DATA_TYPE = 'PL/SQL RECORD' and 1 = 0) then a.PACKAGE_NAME else a.TYPE_NAME end TYPE_NAME, case when a.DATA_TYPE = 'REF CURSOR' then null when (a.DATA_TYPE = 'PL/SQL RECORD' and 1 = 0) then (a.OBJECT_NAME || nvl2(a.ARGUMENT_NAME, ('_' || a.ARGUMENT_NAME), '')) else a.TYPE_SUBNAME end TYPE_SUBNAME, a.POSITION, a.DEFAULTED from SYS.ALL_ARGUMENTS a where (a.OWNER = :1  and a.OBJECT_NAME = :2  and a.PACKAGE_NAME = :3  and decode(a.OVERLOAD, :4 , 1, 0) = 1 and a.DATA_LEVEL = :5  and a.DATA_TYPE is not null) order by a.POSITION asc -- SQL rendered with a free trial version of jOOQ 3.18.0-SNAPSHOT	select a.IN_OUT, a.ARGUMENT_NAME, a.DATA_TYPE, a.CHAR_LENGTH, a.DATA_PRECISION, a.DATA_SCALE, case when a.DATA_TYPE = 'REF CURSOR' then null when (a.DATA_TYPE = 'PL/SQL RECORD' and 1 = 0) then a.OWNER else a.TYPE_OWNER end TYPE_OWNER, case when a.DATA_TYPE = 'REF CURSOR' then null when (a.DATA_TYPE = 'PL/SQL RECORD' and 1 = 0) then a.PACKAGE_NAME else a.TYPE_NAME end TYPE_NAME, case when a.DATA_TYPE = 'REF CURSOR' then null when (a.DATA_TYPE = 'PL/SQL RECORD' and 1 = 0) then (a.OBJECT_NAME || nvl2(a.ARGUMENT_NAME, ('_' || a.ARGUMENT_NAME), '')) else a.TYPE_SUBNAME end TYPE_SUBNAME, a.POSITION, a.DEFAULTED from SYS.ALL_ARGUMENTS a where (a.OWNER = :1  and a.OBJECT_NAME = :2  and a.PACKAGE_NAME = :3  and decode(a.OVERLOAD, :4 , 1, 0) = 1 and a.DATA_LEVEL = :5  and a.DATA_TYPE is not null) order by a.POSITION asc -- SQL rendered with a free trial version of jOOQ 3.18.0-SNAPSHOT
8bkgtn8s08tub	0	3.02141	1	3.021	with c as (select /*+MATERIALIZE*/ c.OWNER, c.TABLE_NAME, c.CONSTRAINT_NAME from SYS.ALL_CONSTRAINTS c where c.OWNER in (:1 , :2 , :3 , :4 )) select SYS.ALL_INDEXES.OWNER, SYS.ALL_INDEXES.INDEX_NAME, SYS.ALL_INDEXES.TABLE_OWNER, SYS.ALL_INDEXES.TABLE_NAME, decode(SYS.ALL_INDEXES.UNIQUENESS, 'UNIQUE', 1, 0) UNIQUENESS from SYS.ALL_INDEXES where (SYS.ALL_INDEXES.OWNER in (:5 , :6 , :7 , :8 ) and SYS.ALL_INDEXES.TABLE_OWNER in (:9 , :10 , :11 , :12 ) and SYS.ALL_INDEXES.INDEX_TYPE = 'NORMAL' and (SYS.ALL_INDEXES.OWNER, SYS.ALL_INDEXES.TABLE_NAME, SYS.ALL_INDEXES.INDEX_NAME) not in ((select c.OWNER, c.TABLE_NAME, c.CONSTRAINT_NAME from c))) order by SYS.ALL_INDEXES.OWNER, SYS.ALL_INDEXES.INDEX_NAME -- SQL rendered with a free trial version of jOOQ 3.18.0-SNAPSHOT	with c as (select /*+MATERIALIZE*/ c.OWNER, c.TABLE_NAME, c.CONSTRAINT_NAME from SYS.ALL_CONSTRAINTS c where c.OWNER in (:1 , :2 , :3 , :4 )) select SYS.ALL_INDEXES.OWNER, SYS.ALL_INDEXES.INDEX_NAME, SYS.ALL_INDEXES.TABLE_OWNER, SYS.ALL_INDEXES.TABLE_NAME, decode(SYS.ALL_INDEXES.UNIQUENESS, 'UNIQUE', 1, 0) UNIQUENESS from SYS.ALL_INDEXES where (SYS.ALL_INDEXES.OWNER in (:5 , :6 , :7 , :8 ) and SYS.ALL_INDEXES.TABLE_OWNER in (:9 , :10 , :11 , :12 ) and SYS.ALL_INDEXES.INDEX_TYPE = 'NORMAL' and (SYS.ALL_INDEXES.OWNER, SYS.ALL_INDEXES.TABLE_NAME, SYS.ALL_INDEXES.INDEX_NAME) not in ((select c.OWNER, c.TABLE_NAME, c.CONSTRAINT_NAME from c))) order by SYS.ALL_INDEXES.OWNER, SYS.ALL_INDEXES.INDEX_NAME -- SQL rendered with a free trial version of jOOQ 3.18.0-SNAPSHOT
2p2w9ju32g2x6	0	2.528278	1	2.528	select SYS.ALL_CONS_COLUMNS.OWNER, SYS.ALL_CONS_COLUMNS.CONSTRAINT_NAME, SYS.ALL_CONS_COLUMNS.TABLE_NAME, SYS.ALL_CONS_COLUMNS.COLUMN_NAME, alias_72650702.R_CONSTRAINT_NAME, alias_72650702.R_OWNER, alias_89229594.TABLE_NAME from (SYS.ALL_CONS_COLUMNS join (SYS.ALL_CONSTRAINTS alias_72650702 left outer join SYS.ALL_CONSTRAINTS alias_89229594 on (alias_72650702.R_OWNER = alias_89229594.OWNER and alias_72650702.R_CONSTRAINT_NAME = alias_89229594.CONSTRAINT_NAME)) on (SYS.ALL_CONS_COLUMNS.OWNER = alias_72650702.OWNER and SYS.ALL_CONS_COLUMNS.CONSTRAINT_NAME = alias_72650702.CONSTRAINT_NAME)) where (alias_72650702.CONSTRAINT_TYPE = 'R' and alias_72650702.OWNER in (:1 , :2 , :3 , :4 )) order by SYS.ALL_CONS_COLUMNS.OWNER, SYS.ALL_CONS_COLUMNS.TABLE_NAME, SYS.ALL_CONS_COLUMNS.CONSTRAINT_NAME, SYS.ALL_CONS_COLUMNS.POSITION -- SQL rendered with a free trial version of jOOQ 3.18.0-SNAPSHOT	select SYS.ALL_CONS_COLUMNS.OWNER, SYS.ALL_CONS_COLUMNS.CONSTRAINT_NAME, SYS.ALL_CONS_COLUMNS.TABLE_NAME, SYS.ALL_CONS_COLUMNS.COLUMN_NAME, alias_72650702.R_CONSTRAINT_NAME, alias_72650702.R_OWNER, alias_89229594.TABLE_NAME from (SYS.ALL_CONS_COLUMNS join (SYS.ALL_CONSTRAINTS alias_72650702 left outer join SYS.ALL_CONSTRAINTS alias_89229594 on (alias_72650702.R_OWNER = alias_89229594.OWNER and alias_72650702.R_CONSTRAINT_NAME = alias_89229594.CONSTRAINT_NAME)) on (SYS.ALL_CONS_COLUMNS.OWNER = alias_72650702.OWNER and SYS.ALL_CONS_COLUMNS.CONSTRAINT_NAME = alias_72650702.CONSTRAINT_NAME)) where (alias_72650702.CONSTRAINT_TYPE = 'R' and alias_72650702.OWNER in (:1 , :2 , :3 , :4 )) order by SYS.ALL_CONS_COLUMNS.OWNER, SYS.ALL_CONS_COLUMNS.TABLE_NAME, SYS.ALL_CONS_COLUMNS.CONSTRAINT_NAME, SYS.ALL_CONS_COLUMNS.POSITION -- SQL rendered with a free trial version of jOOQ 3.18.0-SNAPSHOT
```

The execution plan is:

```
SQL_ID  23mrzz17v4hhq, child number 0
-------------------------------------
select a.IN_OUT, a.ARGUMENT_NAME, a.DATA_TYPE, a.CHAR_LENGTH, 
a.DATA_PRECISION, a.DATA_SCALE, case when a.DATA_TYPE = 'REF CURSOR' 
then null when (a.DATA_TYPE = 'PL/SQL RECORD' and 1 = 0) then a.OWNER 
else a.TYPE_OWNER end TYPE_OWNER, case when a.DATA_TYPE = 'REF CURSOR' 
then null when (a.DATA_TYPE = 'PL/SQL RECORD' and 1 = 0) then 
a.PACKAGE_NAME else a.TYPE_NAME end TYPE_NAME, case when a.DATA_TYPE = 
'REF CURSOR' then null when (a.DATA_TYPE = 'PL/SQL RECORD' and 1 = 0) 
then (a.OBJECT_NAME || nvl2(a.ARGUMENT_NAME, ('_' || a.ARGUMENT_NAME), 
'')) else a.TYPE_SUBNAME end TYPE_SUBNAME, a.POSITION, a.DEFAULTED from 
SYS.ALL_ARGUMENTS a where (a.OWNER = :1  and a.OBJECT_NAME = :2  and 
a.PACKAGE_NAME = :3  and decode(a.OVERLOAD, :4 , 1, 0) = 1 and 
a.DATA_LEVEL = :5  and a.DATA_TYPE is not null) order by a.POSITION asc 
-- SQL rendered with a free trial version of jOOQ 3.18.0-SNAPSHOT
 
Plan hash value: 552951971
 
-------------------------------------------------------------------------------------------------------------------------------------
| Id  | Operation                  | Name              | Starts | E-Rows | A-Rows |   A-Time   | Buffers |  OMem |  1Mem | Used-Mem |
-------------------------------------------------------------------------------------------------------------------------------------
|   0 | SELECT STATEMENT           |                   |      1 |        |      3 |00:00:00.09 |    3389 |       |       |          |
|   1 |  SORT ORDER BY             |                   |      1 |      1 |      3 |00:00:00.09 |    3389 |  2048 |  2048 | 2048  (0)|
|*  2 |   FILTER                   |                   |      1 |        |      3 |00:00:00.09 |    3389 |       |       |          |
|   3 |    PARTITION LIST ALL      |                   |      1 |      1 |      3 |00:00:00.09 |    3389 |       |       |          |
|*  4 |     EXTENDED DATA LINK FULL| INT$DBA_ARGUMENTS |      2 |      1 |      3 |00:00:00.09 |    3389 |       |       |          |
|*  5 |    FIXED TABLE FULL        | X$KZSPR           |      1 |      2 |      1 |00:00:00.01 |       0 |       |       |          |
|   6 |    NESTED LOOPS SEMI       |                   |      0 |      1 |      0 |00:00:00.01 |       0 |       |       |          |
|*  7 |     FIXED TABLE FULL       | X$KZSRO           |      0 |      2 |      0 |00:00:00.01 |       0 |       |       |          |
|*  8 |     INDEX RANGE SCAN       | I_OBJAUTH1        |      0 |      1 |      0 |00:00:00.01 |       0 |       |       |          |
-------------------------------------------------------------------------------------------------------------------------------------
 
Predicate Information (identified by operation id):
---------------------------------------------------
 
   2 - filter(("OWNER"=SYS_CONTEXT('USERENV','CURRENT_USER') OR  IS NOT NULL OR  IS NOT NULL))
   4 - filter(("DATA_TYPE" IS NOT NULL AND "OWNER"=:1 AND "OBJECT_NAME"=:2 AND "PACKAGE_NAME"=:3 AND "DATA_LEVEL"=:5 AND 
              DECODE("OVERLOAD",:4,1,0)=1))
   5 - filter((((-"KZSPRPRV")=(-144) OR (-"KZSPRPRV")=(-141)) AND INTERNAL_FUNCTION("CON_ID") AND 
              "INST_ID"=USERENV('INSTANCE')))
   7 - filter(("CON_ID"=0 OR "CON_ID"=3))
   8 - access("OBJ#"=OBJ_ID(:B1,NVL(:B2,:B3),:B4,:B5) AND "GRANTEE#"="KZSROROL" AND "PRIVILEGE#"=12)
       filter(("GRANTEE#"="KZSROROL" AND "PRIVILEGE#"=12))
```

As can be seen, there are only `filter` predicates, hardly any `access` predicates, for the `(OWNER, OBJECT_NAME, PACKAGE_NAME)` tuple. Hence, this query runs moderately slowly *per execution*. It may be worthwhile querying the entire data set for all input schemata only once, and then filtering by identifier/overload in memory, rather than in the query, given that we have to fetch the same data anyway.

This idea had already been expressed here, in the past: https://github.com/jOOQ/jOOQ/issues/8461

**Comment by lukaseder** [comment]
Yep, definitely another big win!

**Before**

```
11:41:22,252 [INFO] ------------------------------------------------------------------------
11:41:22,257 [INFO] BUILD SUCCESS
11:41:22,258 [INFO] ------------------------------------------------------------------------
11:41:22,262 [INFO] Total time:  01:09 min
11:41:22,264 [INFO] Finished at: 2022-09-23T11:41:22+02:00
11:41:22,266 [INFO] ------------------------------------------------------------------------
```

**After**

```
12:47:42,640 [INFO] ------------------------------------------------------------------------
12:47:42,644 [INFO] BUILD SUCCESS
12:47:42,646 [INFO] ------------------------------------------------------------------------
12:47:42,650 [INFO] Total time:  30.903 s
12:47:42,652 [INFO] Finished at: 2022-09-23T12:47:42+02:00
12:47:42,654 [INFO] ------------------------------------------------------------------------
```

**Comment by lukaseder** [comment]
Curiously, other queries against `ALL_ARGUMENTS` weren't affected, perhaps this is some projection or filter problem, e.g. on `DATA_TYPE`, or whatever.

**Comment by lukaseder** [comment]
Fixed in jOOQ 3.18.0, 3.17.5 (https://github.com/jOOQ/jOOQ/issues/14022), and 3.16.11 (https://github.com/jOOQ/jOOQ/issues/14023)
